storage/client/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -296,35 +331,50 @@ public PageBlobAsyncClient asPageBlobAsyncClient() {
      *         {@link BlobAccessConditions}
      * @param rangeGetContentMD5
      *         Whether the contentMD5 for the specified blob range should be returned.
-     * @param context
-     *         {@code Context} offers a means of passing arbitrary data (key/value pairs) to an
-     *         {@link com.azure.core.http.HttpPipeline}'s policy objects. Most applications do not need to pass
-     *         arbitrary data to the pipeline and can pass {@code Context.NONE} or {@code null}. Each context object is
-     *         immutable. The {@code withContext} with data method creates a new {@code Context} object that refers to
-     *         its parent, forming a linked list.
      */
     public Mono<Void> downloadToFile(String filePath, BlobRange range, BlobAccessConditions accessConditions,
-            boolean rangeGetContentMD5, ReliableDownloadOptions options, Context context) {
-        //todo make this method smart
-        return Mono.using(
-            () -> new FileOutputStream(new File(filePath)),
-            fstream -> this.download(range, accessConditions, rangeGetContentMD5, options, context)
-                .doOnNext(byteBuffer -> {
-                    try {
-                        fstream.write(byteBuffer.array());
-                    } catch (IOException e) {
-                        throw new UncheckedIOException(e);
-                    }
-                })
-                .then(),
-            fstream -> {
+                                     boolean rangeGetContentMD5, ReliableDownloadOptions options) {
+        AsynchronousFileChannel channel;
+        try {
+            channel = AsynchronousFileChannel.open(Paths.get(filePath), StandardOpenOption.READ, StandardOpenOption.WRITE);
+        } catch (IOException e) {
+            return Mono.error(e);
+        }
+        return Mono.justOrEmpty(range)
+            .switchIfEmpty(getFullBlobRange(accessConditions))
+            .flatMapMany(rg -> Flux.fromIterable(sliceBlobRange(rg)))
+            .flatMap(chunk -> blobAsyncRawClient
+                .download(chunk, accessConditions, rangeGetContentMD5)
+                .subscribeOn(Schedulers.elastic())
+                .flatMap(dar -> FluxUtil.bytebufStreamToFile(dar.body(options), channel, chunk.offset() - (range == null ? 0 : range.offset())))
+                .timeout(Duration.ofSeconds(300))
+                .retry(3, throwable -> throwable instanceof IOException || throwable instanceof TimeoutException))
+            .then()
+            .doOnTerminate(() -> {
                 try {
-                    fstream.close();
+                    channel.close();
                 } catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
+            });
+    }
+
+    private Mono<BlobRange> getFullBlobRange(BlobAccessConditions accessConditions) {
+        return getProperties(accessConditions).map(rb -> new BlobRange(0, rb.value().blobSize()));
```

Comment:
We can't make an extra getProperties call just to get the size. This potentially doubles the number of IO operations. Is there a reason why we didn't just pull the implementation from v11?

Reply:
This also does not support etag locking or progress reporting.

Reply:
This is called only once for each large file download, how is it doubling the number of IOs? If the files are really small, then yes - but without the full length the client is not able to parallelize the download work. The best way to avoid it is to provide the `BlobRange` parameter and I'll document it here.

Reply:
I don't understand what is etag locking so feel free to contribute. What kind of progress reporting are you looking for?